### PR TITLE
Add Picard.reorder_sam

### PR DESCRIPTION
**Pull request by ihodes**
for https://github.com/hammerlab/epidisco/issues/167#issuecomment-283806812

### CODE REVIEW

File: src/bfx_tools/picard.ml [hunk 1]
```diff
@@ -150,6 +150,55 @@ let mark_duplicates
...
+    ?reference_build ~input_bam output_bam_path
+  =
+  let open KEDSL in
+  let picard_jar = Machine.get_tool run_with Machine.Tool.Default.picard in
+  let tmp_dir =
+    Filename.chop_extension output_bam_path ^ ".tmpdir" in
+  let input_bam_path = input_bam#product#path in
+  let name =
+    sprintf "picard-reorder_sam-%s" Filename.(basename input_bam#product#path) in
+  let product = transform_bam input_bam#product output_bam_path in
```
Comment (smondet) [inline src/bfx_tools/picard.ml]: In the case where `reference_build` is `Some nrb`, the new bam should carry the new `nrb` as its ref-build I think.

File: src/bfx_tools/picard.ml [hunk 2]
```diff
@@ -150,6 +150,55 @@ let mark_duplicates
...
+          tmp_dir
+          (Filename.quote input_bam_path)
+          (Filename.quote output_bam_path)
+          reference_path)
+    in
+    Machine.run_big_program ~name run_with program
+      ~self_ids:["picard"; "mark-duplicates"] in
+  workflow_node product
+    ~name ~make ~edges:[
+    depends_on input_bam;
```
Comment (smondet) [inline src/bfx_tools/picard.ml]: Do we know whether the tools works on unsorted bams? and indexing?

File: src/pipeline_edsl/semantics.ml [hunk 3]
```diff
@@ -173,6 +173,12 @@ module type Bioinformatics_base = sig
     [ `Bam ] repr ->
     [ `Bam ] repr
 
+  val picard_reorder_sam:
+    ?mem_param : string ->
```
Comment (smondet) [inline src/pipeline_edsl/semantics.ml]: The other tools have a `?configuration` argument which seems cleaner (?)

File: src/pipeline_edsl/to_workflow.ml [hunk 4]
```diff
@@ -731,6 +731,23 @@ module Make (Config : Compiler_configuration)
         ~run_with ~input_bam output_bam
     )
 
+  let picard_reorder_sam ?mem_param ?reference_build bam =
+    let input_bam = get_bam bam in
+    let reference_build_param =
+      match reference_build with
+      | None -> ""
```
Comment (smondet) [inline src/pipeline_edsl/to_workflow.ml]: shouldn't it be `input_bam#ref_build`?

File: src/bfx_tools/picard.ml (hunk 2 above)
Comment (ihodes) [inline src/bfx_tools/picard.ml]: Per @julia326 it should.

File: src/pipeline_edsl/semantics.ml (hunk 3 above)
Comment (ihodes) [inline src/pipeline_edsl/semantics.ml]: I was thinking about it, but for something which has only two params, I thought that would be heavyweight (and easy to introduce later if needed).

**Comment by ihodes** [comment]
@smondet back to you!

**Pull request merged**